src/libraries/Microsoft.Extensions.Logging/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -0,0 +1,661 @@
+                                continue;
+                            }
+
+                            foreach (AttributeListSyntax mal in method.AttributeLists)
+                            {
+                                foreach (AttributeSyntax ma in mal.Attributes)
+                                {
+                                    sm ??= _compilation.GetSemanticModel(classDec.SyntaxTree);
+
+                                    IMethodSymbol attrCtorSymbol = sm.GetSymbolInfo(ma, _cancellationToken).Symbol as IMethodSymbol;
+                                    if (attrCtorSymbol == null || !loggerMessageAttribute.Equals(attrCtorSymbol.ContainingType, SymbolEqualityComparer.Default))
+                                    {
+                                        // badly formed attribute definition, or not the right attribute
+                                        continue;
+                                    }
+
+                                    (int eventId, int? level, string? message, string? eventName) = ExtractAttributeValues(ma.ArgumentList!, sm);
+
+                                    IMethodSymbol? methodSymbol = sm.GetDeclaredSymbol(method, _cancellationToken);
+                                    if (methodSymbol != null)
+                                    {
+                                        var lm = new LoggerMethod
+                                        {
+                                            Name = method.Identifier.ToString(),
+                                            Level = level,
+                                            Message = message ?? string.Empty,
+                                            EventId = eventId,
+                                            EventName = eventName,
+                                            IsExtensionMethod = methodSymbol.IsExtensionMethod,
+                                            Modifiers = method.Modifiers.ToString(),
+                                        };
+
+                                        ExtractTemplates(message, lm.TemplateMap, lm.TemplateList);
+
+                                        bool keepMethod = true;   // whether or not we want to keep the method definition or if it's got errors making it so we should discard it instead
+                                        if (lm.Name[0] == '_')
+                                        {
+                                            // can't have logging method names that start with _ since that can lead to conflicting symbol names
+                                            // because the generated symbols start with _
+                                            Diag(DiagnosticDescriptors.InvalidLoggingMethodName, method.Identifier.GetLocation());
+                                        }
+
+                                        if (sm.GetTypeInfo(method.ReturnType!).Type!.SpecialType != SpecialType.System_Void)
```

Comment:
You already got the `IMethodSymbol` from the semantic model, it would be less expensive if you just did, `methodSymbol.ReturnsVoid`.

Reply:
Feel free to address on a follow up PR.